packages/context/src/context-event-notifier.ts

```diff
@@ -0,0 +1,189 @@
+// Copyright IBM Corp. 2017,2018. All Rights Reserved.
+// Node module: @loopback/context
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {EventEmitter} from 'events';
+import {
+  ContextEventObserver,
+  ContextEventType,
+  Notification,
+  Subscription,
+  ContextObserver,
+} from './context-observer';
+import {Context} from './context';
+import {ContextSubscription} from './context-subscription';
+import {Binding} from './binding';
+
+// FIXME: `@types/p-event` is out of date against `p-event@2.2.0`
+const pEvent = require('p-event');
+
+export class ContextEventNotifier extends EventEmitter {
+  /**
+   * A list of registered context observers. The Set will be created when the
+   * first observer is added.
+   */
+  protected observers: Set<ContextEventObserver> | undefined;
+
+  /**
+   * Internal counter for pending notification events which are yet to be
+   * processed by observers.
+   */
+  private pendingNotifications = 0;
+
+  /**
+   * Queue for background notifications for observers
+   */
+  private notificationQueue: AsyncIterableIterator<Notification> | undefined;
+
+  constructor(
+    // tslint:disable-next-line:no-any
+    private _debug: (...args: any[]) => void,
```

Comment:
The constructor is a bit ugly. We can get rid of it as the `Context` object is available for the two usages of this function. For example:
```ts
await this.notifyObservers(eventType, binding, context, observers);
        this.pendingNotifications--;
        debug(
          '[%s] Observers notified for %s of binding %s',
          eventType,
          binding.key,
          context.name,
        );
```

Reply:
BTW, it might be desirable to change the constructor to be:
```ts
constructor(private context: Context) {}
```